## Project 6: Matrix Methods

### Abstract & Description:

This project involves performing thick lens calculations using a matrix representation of optical elements. In this experiment, we obtained the following results:

System Matrix: $$M = \begin{bmatrix} 1 &3 \\-0.24752475 & 0.25742574  \end{bmatrix} $$
 and also obtained the following as focal length:
 $$ f = 4.04$$
### Algorithm & Discussion: 

The math behind all this is quite long. The good thing is that it all boils down to two main matrix operators: Refraction and Translation.

#### Refraction:
$$\hat{R}\left|r_0\right\rangle = \begin{bmatrix} 1 &amp; 0 \\ \frac{n_l - n_r}{R n_r} &amp; \frac{n_l}{n_r} \end{bmatrix} \begin{bmatrix} y_0 \\ \alpha_0\end{bmatrix}$$
#### Translation
$$\hat{T}\left|r_0\right\rangle = \begin{bmatrix} 1 &amp; L \\ 0 &amp; 1 \end{bmatrix} \begin{bmatrix} y_0 \\ \alpha_0\end{bmatrix}$$

Where $n_l$ and $n_r$ are the indices of refraction on the left and right (respectively) of the refractive interface. $R$ is the radius of curvature of the refractive interface (positive if the center of curvature is to the right, negative to the left). $y_0$ is the height of the ray before the transition, $\alpha_0$ is the angle (in radians) of the ray before the transition. $L$ is the length of the translation. Various combinations of these two operators can describe very complicated lens systems easily!

### Implementation/ Code:


In [28]:
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [29]:
ng=1.5  # index of glass
na=1.0  # index of air
h=0.8  # start with ray 1.0 cm above axis
R=2.02  # radius of curvature of lens

ray1 = array([[h],[0.0]])

ray1

array([[0.8],
       [0. ]])

In [30]:
T1 = array([[1.0, 3.0],
            [0.0, 1.0]
            ])
T1

array([[1., 3.],
       [0., 1.]])

In [31]:
ray2 = T1.dot(ray1)  # matrix multiplication is handled by the "dot" method of an array
ray2

array([[0.8],
       [0. ]])

In [32]:
R1 = array([[1.0, 0.0],               # entering the curved surface
            [(na-ng)/(ng*R), na/ng]])
R1

array([[ 1.        ,  0.        ],
       [-0.1650165 ,  0.66666667]])

In [33]:
ray3 = R1.dot(ray2)
ray3

array([[ 0.8      ],
       [-0.1320132]])

In [34]:
R2 = array([[1.0,0.0],              # exiting the planer surface
            [0.0, ng/na]])
R2

array([[1. , 0. ],
       [0. , 1.5]])

In [35]:
ray4=R2.dot(ray3)
ray4

array([[ 0.8      ],
       [-0.1980198]])

In [36]:
fl=-ray4[0,0]/ray4[1,0]         # calculate the focal length from the height and angle of the ray.
fl

4.04

In [37]:
na*R/(ng-na)                    # compare to the "lens makers" equation result.


4.04

In [40]:
M = R2.dot(R1.dot(T1))   # system matrix
M

array([[ 1.        ,  3.        ],
       [-0.24752475,  0.25742574]])

In [41]:
M.dot(ray1)             # system acting on ray1

array([[ 0.8      ],
       [-0.1980198]])

### Results

Using the actual math, in order to solve for for the focal length, we need to use the following formula:
 $$f = \frac{n_gxR}{n_g-n_a} = 4.04$$
What did the experiment give us?
We obtained the same exact answer with absolutely no error within the 2 first decimal digits. 
After applying the Refraction and Translation matrix operator, we ended up with the following:

 $$M = \begin{bmatrix} 1 &3; 0 \\-0.24752475 & 0.25742574  \end{bmatrix} $$

The system matrix is very important when it comes to solving for different systems acting on given rays.
As an illustration, by doing the matrix multiplication of the system matrix and ray1's entries, we ended up with the following:
  $$M = \begin{bmatrix} 0.8; 0 \\-0.1980198  \end{bmatrix} $$
  
### Conclusion

For this experiment, we were trying to use Matrices in order to solve Optics problems. This was merely for the sake of familiarizing students with the use of matrices in python and also as a tool for solving scientific problems. Altogether, we solved for the focal length and obtained a solution with no error. 